File: python/src/anysearch/providers/parallel.py
"""Parallel (parallel.ai) — LLM-optimized web search returning ranked excerpts."""

from __future__ import annotations

import httpx

from .._http import PreparedRequest
from ..types import SearchRequest, SearchResponse
from .base import BaseProvider, Capability

_MODE = {"fast": "basic", "balanced": "basic", "deep": "advanced"}


class ParallelProvider(BaseProvider):
    name = "parallel"
    env_keys = ("PARALLEL_API_KEY",)
    default_base_url = "https://api.parallel.ai"
    extra_package = "parallel-web"
    native_client = ("parallel", ("Parallel",))
    capabilities = frozenset(
        {
            Capability.DOMAINS,
            Capability.COUNTRY,
            Capability.DATE,
            Capability.MODE,
            Capability.CONTENT,
            Capability.HIGHLIGHTS,
        }
    )

    def prepare(self, req: SearchRequest) -> PreparedRequest:
        body = {
            "search_queries": [req.query],
            "objective": req.extra.pop("objective", req.query),
            "max_results": req.max_results,
        }
        if req.mode:
            body["mode"] = _MODE.get(req.mode, "advanced")

        advanced = {}
        source_policy = {}
        if req.include_domains:
            source_policy["include_domains"] = req.include_domains
        if req.exclude_domains:
            source_policy["exclude_domains"] = req.exclude_domains
        if req.start_published_date:
            source_policy["after_date"] = req.start_published_date
        if source_policy:
            advanced["source_policy"] = source_policy
        if req.country:
            advanced["location"] = req.country.lower()
        if advanced:
            body["advanced_settings"] = advanced

        body.update(req.extra)
        return PreparedRequest(
            method="POST",
            url=f"{self.base_url}/v1/search",
            headers={"x-api-key": self.api_key, "Content-Type": "application/json"},
            json=body,
        )

    def parse(self, response: httpx.Response, req: SearchRequest, elapsed_ms: float) -> SearchResponse:
        data = response.json()
        results = []
        for item in data.get("results", []) or []:
            excerpts = item.get("excerpts") or []
            if isinstance(excerpts, str):
                excerpts = [excerpts]
            text = "\n\n".join(e for e in excerpts if e) or None
            results.append(
                self._result(
                    title=item.get("title"),
                    url=item.get("url"),
                    snippet=(excerpts[0] if excerpts else None),
                    text=text,
                    highlights=list(excerpts),
                    raw=item,
                )
            )
        return self._response(
            req, results, data, request_id=data.get("search_id"), elapsed_ms=elapsed_ms
        )

File: js/src/providers/parallel.ts
import type { HttpResponse } from "../http.js";
import type { SearchRequest, SearchResponse } from "../types.js";
import { finalize, result, type ProviderContext, type ProviderSpec } from "./base.js";

const MODE: Record<string, string> = { fast: "basic", balanced: "basic", deep: "advanced" };

export const parallel: ProviderSpec = {
  name: "parallel",
  envKeys: ["PARALLEL_API_KEY"],
  defaultBaseUrl: "https://api.parallel.ai",
  extraPackage: "parallel-web",
  nativeImport: { module: "parallel-web", exports: ["Parallel", "default"] },
  capabilities: new Set(["domains", "country", "date", "mode", "content", "highlights"]),

  prepare(ctx: ProviderContext, req: SearchRequest) {
    const extra = { ...req.extra };
    const objective = (extra.objective as string) ?? req.query;
    delete extra.objective;
    const body: Record<string, unknown> = {
      search_queries: [req.query],
      objective,
      max_results: req.maxResults,
    };
    if (req.mode) body.mode = MODE[req.mode] ?? "advanced";

    const sourcePolicy: Record<string, unknown> = {};
    if (req.includeDomains.length) sourcePolicy.include_domains = req.includeDomains;
    if (req.excludeDomains.length) sourcePolicy.exclude_domains = req.excludeDomains;
    if (req.startPublishedDate) sourcePolicy.after_date = req.startPublishedDate;
    const advanced: Record<string, unknown> = {};
    if (Object.keys(sourcePolicy).length) advanced.source_policy = sourcePolicy;
    if (req.country) advanced.location = req.country.toLowerCase();
    if (Object.keys(advanced).length) body.advanced_settings = advanced;

    Object.assign(body, extra);
    return {
      method: "POST",
      url: `${ctx.baseUrl}/v1/search`,
      headers: { "x-api-key": ctx.apiKey ?? "", "Content-Type": "application/json" },
      json: body,
    };
  },

  parse(_ctx, res: HttpResponse, req: SearchRequest, elapsedMs: number): SearchResponse {
    const data = res.data ?? {};
    const results = (data.results ?? []).map((item: any) => {
      const excerpts: string[] = Array.isArray(item.excerpts)
        ? item.excerpts
        : item.excerpts
          ? [item.excerpts]
          : [];
      return result({
        title: item.title,
        url: item.url,
        snippet: excerpts[0],
        text: excerpts.filter(Boolean).join("\n\n") || undefined,
        highlights: excerpts,
        raw: item,
      });
    });
    return finalize(parallel, req, results, data, { requestId: data.search_id, elapsedMs });
  },
};
